File: example/ios/Runner/DeepLinkPlugin.swift
import Flutter
import UIKit
import Foundation

/**
 * DeepLinkPlugin handles deep link processing for MRTD validation on iOS
 * Integrates with Flutter's MethodChannel for cross-platform communication
 */
@objc class DeepLinkPlugin: NSObject, FlutterPlugin {
    private static let channelName = "deep_link_handler"
    private static let methodGetInitialLink = "getInitialLink"
    private static let methodHandleDeepLink = "handleDeepLink"
    
    private var channel: FlutterMethodChannel?
    private var initialLink: String?
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
        let instance = DeepLinkPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case DeepLinkPlugin.methodGetInitialLink:
            result(initialLink)
            initialLink = nil // Clear after first access
            
        case DeepLinkPlugin.methodHandleDeepLink:
            guard let url = call.arguments as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "URL is required", details: nil))
                return
            }
            let handled = processDeepLink(url: url)
            result(["success": handled])
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension DeepLinkPlugin {
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Handle launch via URL scheme (custom schemes)
        if let url = launchOptions?[.url] as? URL {
            handleURL(url)
        }
        return true
    }
    
    func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
        return handleURL(url)
    }
    
    // iOS 13+ Scene Delegate support (Universal Links)
    func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
        if userActivity.activityType == NSUserActivityTypeBrowsingWeb,
           let url = userActivity.webpageURL {
            return handleURL(url)
        }
        return false
    }
}

private extension DeepLinkPlugin {
    
    func handleURL(_ url: URL) -> Bool {
        let urlString = url.absoluteString
        
        guard isValidURL(url) else {
            NSLog("Invalid URL: \(urlString)")
            return false
        }
        
        // If Flutter is ready, send immediately
        if let channel = channel {
            channel.invokeMethod(DeepLinkPlugin.methodHandleDeepLink, arguments: urlString)
        } else {
            // Store for later retrieval
            initialLink = urlString
        }
        
        return true
    }
    
    func isValidURL(_ url: URL) -> Bool {
        // Require HTTPS
        guard url.scheme == "https" else { return false }
        
        // Require host/path to match start-app endpoint
        let host = url.host ?? ""
        let path = url.path
        guard host == "passport-issuer.staging.yivi.app",
              path.hasPrefix("/start-app") else { return false }
        
        // Extract query parameters
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let queryItems = components.queryItems,
              !queryItems.isEmpty else { return false }
        
        var params: [String: String] = [:]
        for item in queryItems {
            if let value = item.value { params[item.name] = value }
        }
        
        // Validate required params
        guard let sessionId = params["sessionId"], isValidSessionId(sessionId),
              let nonce = params["nonce"], isValidNonce(nonce) else { return false }
        
        return true
    }
    
    func processDeepLink(url: String) -> Bool {
        guard let urlObj = URL(string: url) else { return false }
        guard isValidURL(urlObj) else { return false }
        
        // Log security event
        NSLog("Processing valid MRTD deep link: \(urlObj.host ?? "unknown")")
        return true
    }
    
    // MARK: - Validation Helpers
    
    /// 32-char alphanumeric (not a UUID)
    func isValidSessionId(_ sessionId: String) -> Bool {
        return sessionId.range(of: "^[A-Za-z0-9]{32}$", options: .regularExpression) != nil
    }
    
    /// 16-char hexadecimal
    func isValidNonce(_ nonce: String) -> Bool {
        return nonce.range(of: "^[0-9a-fA-F]{16}$", options: .regularExpression) != nil
    }
}

File: example/android/app/src/main/kotlin/app/yivi/vcmrtd/DeepLinkPlugin.kt
package foundation.privacybydesign.vcmrtd

import android.content.Intent
import android.net.Uri
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.android.FlutterActivity

class DeepLinkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    companion object {
        private const val CHANNEL = "deep_link_handler"
        private const val METHOD_GET_INITIAL_LINK = "getInitialLink"
        private const val METHOD_HANDLE_DEEP_LINK = "handleDeepLink"
    }

    private lateinit var channel: MethodChannel
    private var activity: FlutterActivity? = null
    private var initialLink: String? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as? FlutterActivity
        activity?.intent?.let { handleIntent(it) }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity as? FlutterActivity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            METHOD_GET_INITIAL_LINK -> {
                result.success(initialLink)
                initialLink = null
            }
            METHOD_HANDLE_DEEP_LINK -> {
                val url = call.arguments as? String
                if (url != null) {
                    val handled = processDeepLink(url)
                    result.success(mapOf("success" to handled))
                } else {
                    result.error("INVALID_ARGUMENTS", "URL is required", null)
                }
            }
            else -> result.notImplemented()
        }
    }

    fun onNewIntent(intent: Intent) {
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        if (intent.action == Intent.ACTION_VIEW) {
            val uri = intent.data ?: return
            val url = uri.toString()

            if (isValidPassportIssuerUrl(uri)) {
                if (::channel.isInitialized) {
                    channel.invokeMethod(METHOD_HANDLE_DEEP_LINK, url)
                } else {
                    initialLink = url
                }
            }
        }
    }

    private fun isValidPassportIssuerUrl(uri: Uri): Boolean {
        if (uri.scheme != "https") return false
        if (uri.host != "passport-issuer.staging.yivi.app") return false
        if (!uri.path.orEmpty().startsWith("/start-app")) return false

        val sessionId = uri.getQueryParameter("sessionId")
        val nonce = uri.getQueryParameter("nonce")

        return !sessionId.isNullOrEmpty() && !nonce.isNullOrEmpty()
    }

    private fun processDeepLink(url: String): Boolean {
        return try {
            val uri = Uri.parse(url)
            if (!isValidPassportIssuerUrl(uri)) {
                false
            } else {
                android.util.Log.i("DeepLinkPlugin", "Processing valid deep link: $url")
                true
            }
        } catch (e: Exception) {
            android.util.Log.e("DeepLinkPlugin", "Error processing deep link: ${e.message}")
            false
        }
    }
}
